handler/src/main/java/io/netty/handler/internal/svm/ssl/SslContextSubstitutions.java

```diff
@@ -0,0 +1,78 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.internal.svm.ssl;
+
+import io.netty.handler.ssl.ApplicationProtocolConfig;
+import io.netty.handler.ssl.CipherSuiteFilter;
+import io.netty.handler.ssl.ClientAuth;
+import io.netty.handler.ssl.SslContext;
+import io.netty.handler.ssl.SslProvider;
+
+import java.security.PrivateKey;
+import java.security.Provider;
+import java.security.cert.X509Certificate;
+
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.SSLException;
+import javax.net.ssl.TrustManagerFactory;
+
+import com.oracle.svm.core.annotate.Substitute;
+import com.oracle.svm.core.annotate.TargetClass;
+
+@TargetClass(className = "io.netty.handler.ssl.SslContext")
+final class SslContextSubstitutions {
+
+    private SslContextSubstitutions() {
+    }
+
+    @Substitute
+    static SslContext newServerContextInternal(SslProvider provider,
+        Provider sslContextProvider,
+        X509Certificate[] trustCertCollection, TrustManagerFactory trustManagerFactory,
+        X509Certificate[] keyCertChain, PrivateKey key, String keyPassword, KeyManagerFactory keyManagerFactory,
+        Iterable<String> ciphers, CipherSuiteFilter cipherFilter, ApplicationProtocolConfig apn,
+        long sessionCacheSize, long sessionTimeout, ClientAuth clientAuth, String[] protocols, boolean startTls,
+        boolean enableOcsp, String keyStoreType)
+        throws SSLException {
+
+        if (enableOcsp) {
+            throw new IllegalArgumentException("OCSP is not supported with this SslProvider: " + provider);
+        }
+        return (SslContext) (Object) new JdkSslServerContextSubstitution(
+            sslContextProvider,
+            trustCertCollection, trustManagerFactory, keyCertChain, key, keyPassword,
+            keyManagerFactory, ciphers, cipherFilter, apn, sessionCacheSize, sessionTimeout,
+            clientAuth, protocols, startTls, keyStoreType);
+    }
+
+    @Substitute
+    static SslContext newClientContextInternal(
+        SslProvider provider,
+        Provider sslContextProvider,
+        X509Certificate[] trustCert, TrustManagerFactory trustManagerFactory,
+        X509Certificate[] keyCertChain, PrivateKey key, String keyPassword, KeyManagerFactory keyManagerFactory,
+        Iterable<String> ciphers, CipherSuiteFilter cipherFilter, ApplicationProtocolConfig apn, String[] protocols,
+        long sessionCacheSize, long sessionTimeout, boolean enableOcsp, String keyStoreType) throws SSLException {
+        if (enableOcsp) {
+            throw new IllegalArgumentException("OCSP is not supported with this SslProvider: " + provider);
+        }
+        return (SslContext) (Object) new JdkSslClientContextSubstitution(
```

Comment:
why do we need the double cast ?

Reply:
The make the Java compiler happy. I don't know how Substrate does it internally, but `JdkSslServerContextSubstitution` isn't in the hierarchy of `SslContext` and cannot be casted directly.